### PR TITLE
[Doc] Clarify the functionality of forward_hook.

### LINKED ISSUE #74024: Ambiguous docstring on `register_module_forward_hook`
### 📚 The doc issue

In current docstring of `forward_hook`, it says:

https://github.com/pytorch/pytorch/blob/38b94f41590d8361fd26bd18ec94938faa648a90/torch/nn/modules/module.py#L100-L102

What's the meaning of `It can modify the input inplace but it will not have effect on forward` while `The hook can modify the output.` also holds?

When I check the implementation of `Module`, the `result` of `forward` is indeed overridden by `hook_result` as long as the hook function returns a modified result.

https://github.com/pytorch/pytorch/blob/38b94f41590d8361fd26bd18ec94938faa648a90/torch/nn/modules/module.py#L1130-L1134

### Suggest a potential alternative/fix

I suppose the above docstring to be modified as follows:

```
     The hook can modify the module output as long as it returns 
     the modified output. It cannot modify the given argument `output` 
     inplace since this is called after :func:`forward` is called. 
```

cc @brianjo @mruberry @albanD @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by BrandoZhang**
Fixes #74024.

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -96,9 +96,8 @@ def register_module_forward_hook(hook: Callable[..., None]) -> RemovableHandle:
 
     The input contains only the positional arguments given to the module.
     Keyword arguments won't be passed to the hooks and only to the ``forward``.
-    The hook can modify the output. It can modify the input inplace but
-    it will not have effect on forward since this is called after
-    :func:`forward` is called.
+    You can optionally modify the output of the module by returning a new value
+    that will be replacing the output from the :func:`forward` function.
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:100]: tiny grammar nit: be replacing -> replace

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1067,10 +1066,10 @@ def register_forward_hook(self, hook: Callable[..., None]) -> RemovableHandle:
             hook(module, input, output) -> None or modified output
 
         The input contains only the positional arguments given to the module.
-        Keyword arguments won't be passed to the hooks and only to the ``forward``.
-        The hook can modify the output. It can modify the input inplace but
-        it will not have effect on forward since this is called after
-        :func:`forward` is called.
+        Keyword arguments won't be passed to the hooks and only to the
+        ``forward``. You can optionally modify the output of the module by
+        returning a new value that will be replacing the output from the
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1071]: tiny grammar nit: be replacing -> replace

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Added one small comment on wording but otherwise LGTM :)

**Review by albanD — COMMENTED** [review]
LGTM beyond Joel's comments

**Comment by jbschlosser** [comment]
@BrandoZhang still interested in getting this merged?

**Pull request closed without merging**